syntax = "proto3";

import "users.proto";


service Polls {
    // retorna todas as enquetes
    rpc GetPolls(Empty) returns (GetPollsReply) {} 
    
    // cria uma enquete
    rpc CreatePoll(PollRequest) returns (Poll) {}

    // delete uma enquete
    rpc DeletePoll(PollRequest) returns (Empty) {}
    
    // Vota em uma enquete.
    rpc Vote(VoteInfo) returns (Empty) {}
    
    // remove um voto de uma enquete.
    rpc Unvote(VoteInfo) returns (Empty) {}
    
    // retorna todas as enquetes nas quais um usuário votou
    rpc GetPollsVotedByUser(User) returns (GetPollsReply) {}
    
    // retorna todas as enquetes criadas por um usuário.
    rpc GetUserPolls(User) returns (GetPollsReply) {}

    // retorna o poll com id especifico.
    rpc GetPollID(Poll) returns (Poll) {}
}

message VoteInfo {
    int32 id_user = 1;
    int32 id_option = 2;
}

message PollOptions {
    optional int32 id = 1;
    string text = 2;
    optional int32 votes = 3;
}

message Poll {
    optional int32 id = 1;
    string title = 2;
    string text = 3;
    repeated PollOptions options = 4;
}

message GetPollsReply {
    // mapeia um ID à uma Poll.
    repeated Poll polls = 1;
}

message PollRequest {
    User user = 1;
    Poll poll = 2;
}

